### ISSUE TITLE
Javadoc build failure on JDK 15

### ISSUE THREAD

**Issue by corona10** [line/armeria#3359]
When building javadoc of armeria `./gradlew javadoc:javadoc` on JDK15 environment,
I meet Javadoc build failure.

```
/Users/user/line/armeria/consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListenerBuilder.java:38: error: heading used out of sequence: <H3>, compared to implicit preceding heading: <H1>
 * <h3>Examples</h3>
   ^
/Users/user/line/armeria/consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroupBuilder.java:35: error: heading used out of sequence: <H3>, compared to implicit preceding heading: <H1>
 * <h3>Examples</h3>
```

AFAIK, the Armeria project is now based on JDK 14, so this issue is not a high-priority issue.
But since JDK 14 is not an LTS version so in near future IMHO this issue is should be solved.
And also please note that sdkman does not provide adopt JDK14 version already.

````
➜  ~ sdk list java
================================================================================
Available Java Versions
================================================================================
 Vendor        | Use | Version      | Dist    | Status     | Identifier
--------------------------------------------------------------------------------
 AdoptOpenJDK  |     | 15.0.2.j9    | adpt    |            | 15.0.2.j9-adpt
               |     | 15.0.2.hs    | adpt    | installed  | 15.0.2.hs-adpt
               |     | 11.0.10.j9   | adpt    |            | 11.0.10.j9-adpt
               |     | 11.0.10.hs   | adpt    |            | 11.0.10.hs-adpt
               |     | 11.0.9.hs    | adpt    | local only | 11.0.9.hs-adpt
               |     | 8.0.282.j9   | adpt    |            | 8.0.282.j9-adpt
               |     | 8.0.282.hs   | adpt    |            | 8.0.282.hs-adpt
 Amazon        |     | 15.0.2.7.1   | amzn    |            | 15.0.2.7.1-amzn
               |     | 15.0.1       | amzn    |            | 15.0.1-amzn
               |     | 15.0.1.9.1   | amzn    |            | 15.0.1.9.1-amzn
               |     | 11.0.10.9.1  | amzn    |            | 11.0.10.9.1-amzn
               |     | 8.282.08.1   | amzn    |            | 8.282.08.1-amzn
 Azul Zulu     |     | 15.0.2       | zulu    |            | 15.0.2-zulu
               |     | 15.0.2.fx    | zulu    |            | 15.0.2.fx-zulu
               |     | 15.0.1       | zulu    |            | 15.0.1-zulu
               |     | 15.0.1.fx    | zulu    |            | 15.0.1.fx-zulu
               | >>> | 14.0.2.fx    | zulu    | installed  | 14.0.2.fx-zulu
               |     | 13.0.5.fx    | zulu    |            | 13.0.5.fx-zulu
               |     | 11.0.10      | zulu    |            | 11.0.10-zulu
               |     | 11.0.10.fx   | zulu    |            | 11.0.10.fx-zulu
               |     | 11.0.9       | zulu    |            | 11.0.9-zulu
               |     | 11.0.9.fx    | zulu    |            | 11.0.9.fx-zulu
               |     | 8.0.282      | zulu    |            | 8.0.282-zulu
               |     | 8.0.282.fx   | zulu    |            | 8.0.282.fx-zulu
               |     | 8.0.275.fx   | zulu    |            | 8.0.275.fx-zulu
               |     | 7.0.282      | zulu    |            | 7.0.282-zulu
 BellSoft      |     | 15.0.2.fx    | librca  |            | 15.0.2.fx-librca
               |     | 15.0.2       | librca  |            | 15.0.2-librca
               |     | 11.0.10.fx   | librca  |            | 11.0.10.fx-librca
               |     | 11.0.10      | librca  |            | 11.0.10-librca
               |     | 8.0.282.fx   | librca  |            | 8.0.282.fx-librca
               |     | 8.0.282      | librca  |            | 8.0.282-librca
 GraalVM       |     | 21.0.0.r11   | grl     | local only | 21.0.0.r11-grl
               |     | 21.0.0.2.r11 | grl     |            | 21.0.0.2.r11-grl
               |     | 21.0.0.2.r8  | grl     |            | 21.0.0.2.r8-grl
               |     | 20.3.1.2.r11 | grl     |            | 20.3.1.2.r11-grl
               |     | 20.3.1.2.r8  | grl     |            | 20.3.1.2.r8-grl
               |     | 19.3.5.r11   | grl     |            | 19.3.5.r11-grl
               |     | 19.3.5.r8    | grl     |            | 19.3.5.r8-grl
 Java.net      |     | 17.ea.10     | open    |            | 17.ea.10-open
               |     | 17.ea.9      | open    |            | 17.ea.9-open
               |     | 17.ea.2.pma  | open    |            | 17.ea.2.pma-open
               |     | 17.ea.2.lm   | open    |            | 17.ea.2.lm-open
               |     | 17.ea.1.pma  | open    |            | 17.ea.1.pma-open
               |     | 16.ea.36     | open    |            | 16.ea.36-open
               |     | 15.0.2       | open    |            | 15.0.2-open
               |     | 11.0.2       | open    |            | 11.0.2-open
 SAP           |     | 15.0.2       | sapmchn |            | 15.0.2-sapmchn
               |     | 11.0.10      | sapmchn |            | 11.0.10-sapmchn
 TravaOpenJDK  |     | 11.0.9       | trava   |            | 11.0.9-trava
               |     | 8.0.232      | trava   |            | 8.0.232-trava
================================================================================
Use the Identifier for installation:

    $ sdk install java 11.0.3.hs-adpt
================================================================================
````

Related issue: https://bugs.openjdk.java.net/browse/JDK-8223552

**Comment by corona10** [comment]
From 3cbdcdc081f1735cb2eb660fae6854dad75b818f Mon Sep 17 00:00:00 2001
From: donghee-na <donghee.na@linecorp.com>
Date: Sun, 21 Feb 2021 14:15:54 +0900
Subject: [PATCH] Update javadoc for JDK 15

---
 .../client/consul/ConsulEndpointGroupBuilder.java      |  2 +-
 .../server/consul/ConsulUpdatingListenerBuilder.java   |  2 +-
 .../armeria/common/ContentDispositionBuilder.java      |  2 +-
 .../linecorp/armeria/common/RequestHeaderGetters.java  |  6 ++----
 .../resources/com/linecorp/armeria/public_suffixes.txt | 10 ++++++++--
 .../protobuf/ProtobufRequestConverterFunction.java     |  2 +-
 .../protobuf/ProtobufResponseConverterFunction.java    |  2 +-
 .../armeria/spring/HealthCheckServiceConfigurator.java |  2 +-
 8 files changed, 16 insertions(+), 12 deletions(-)

diff --git a/consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroupBuilder.java b/consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroupBuilder.java
index 4f2fb0036..75cff2f7b 100644
--- a/consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroupBuilder.java
+++ b/consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroupBuilder.java
@@ -32,7 +32,7 @@ import com.linecorp.armeria.server.consul.ConsulUpdatingListenerBuilder;
 
 /**
  * A builder class for {@link ConsulEndpointGroup}.
- * <h3>Examples</h3>
+ * <h2>Examples</h2>
  * <pre>{@code
  * ConsulEndpointGroup endpointGroup = ConsulEndpointGroup.builder(consulUri, "myService")
  *                                                        .build();
diff --git a/consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListenerBuilder.java b/consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListenerBuilder.java
index 01e832fb0..7095dbbe1 100644
--- a/consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListenerBuilder.java
+++ b/consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListenerBuilder.java
@@ -35,7 +35,7 @@ import com.linecorp.armeria.server.Server;
 
 /**
  * Builds a new {@link ConsulUpdatingListener}, which registers the server to Consul cluster.
- * <h3>Examples</h3>
+ * <h2>Examples</h2>
  * <pre>{@code
  * ConsulUpdatingListener listener = ConsulUpdatingListener.builder(consulUri, "myService")
  *                                                         .build();
diff --git a/core/src/main/java/com/linecorp/armeria/common/ContentDispositionBuilder.java b/core/src/main/java/com/linecorp/armeria/common/ContentDispositionBuilder.java
index b2d0a2d73..974cf5eea 100644
--- a/core/src/main/java/com/linecorp/armeria/common/ContentDispositionBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/common/ContentDispositionBuilder.java
@@ -86,7 +86,7 @@ public final class ContentDispositionBuilder {
      * Only the US-ASCII, UTF-8 and ISO-8859-1 charsets are supported.
      *
      * <p><strong>Note:</strong> Do not use this for a {@code "multipart/form-data"} requests as per
-     * <a link="https://tools.ietf.org/html/rfc7578#section-4.2">RFC 7578, Section 4.2</a>
+     * <a href="https://tools.ietf.org/html/rfc7578#section-4.2">RFC 7578, Section 4.2</a>
      * and also RFC 5987 itself mentions it does not apply to multipart requests.
      */
     public ContentDispositionBuilder filename(String filename, @Nullable Charset charset) {
diff --git a/core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java b/core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java
index 6c01a0409..8a89185f5 100644
--- a/core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java
+++ b/core/src/main/java/com/linecorp/armeria/common/RequestHeaderGetters.java
@@ -87,8 +87,7 @@ interface RequestHeaderGetters extends HttpHeaderGetters {
      * algorithm described in
      * <a href="https://tools.ietf.org/html/rfc2616#section-14.4">RFC2616 Accept-Language (obsoleted)</a>
      * and also referenced in <a href="https://tools.ietf.org/html/rfc7231#section-5.3.5">RFC7231 Accept-Language</a>.
-     * <p/>
-     * See also {@link Locale#lookup} for another algorithm.
+     * <p>See also {@link Locale#lookup} for another algorithm.</p>
      * @param supportedLocales a {@link Iterable} of {@link Locale}s supported by the server.
      * @return The best matching {@link Locale} or {@code null} if no locale matches.
      */
@@ -103,8 +102,7 @@ interface RequestHeaderGetters extends HttpHeaderGetters {
      * algorithm described in
      * <a href="https://tools.ietf.org/html/rfc2616#section-14.4">RFC2616 Accept-Language (obsoleted)</a>
      * and also referenced in <a href="https://tools.ietf.org/html/rfc7231#section-5.3.5">RFC7231 Accept-Language</a>.
-     * <p/>
-     * See also {@link Locale#lookup} for another algorithm.
+     * <p>See also {@link Locale#lookup} for another algorithm.</p>
      * @param supportedLocales {@link Locale}s supported by the server.
      * @return The best matching {@link Locale} or {@code null} if no locale matches.
      */
diff --git a/core/src/main/resources/com/linecorp/armeria/public_suffixes.txt b/core/src/main/resources/com/linecorp/armeria/public_suffixes.txt
index c5e4d0c35..08eb014be 100644
--- a/core/src/main/resources/com/linecorp/armeria/public_suffixes.txt
+++ b/core/src/main/resources/com/linecorp/armeria/public_suffixes.txt
@@ -18,6 +18,7 @@
 *.ck
 *.cloud.metacentrum.cz
 *.cns.joyent.com
+*.code.run
 *.compute-1.amazonaws.com
 *.compute.amazonaws.com
 *.compute.amazonaws.com.cn
@@ -54,6 +55,7 @@
 *.nagoya.jp
 *.nodebalancer.linode.com
 *.nom.br
+*.northflank.app
 *.np
 *.oci.customer-oci.com
 *.ocp.customer-oci.com
@@ -607,6 +609,7 @@ avoues.fr
 aw
 awaji.hyogo.jp
 aws
+awsglobalaccelerator.com
 awsmppl.com
 ax
 axa
@@ -824,6 +827,7 @@ biz.ki
 biz.ls
 biz.mv
 biz.mw
+biz.my
 biz.ni
 biz.nr
 biz.pk
@@ -1131,7 +1135,6 @@ casa
 casacam.net
 casadelamoneda.museum
 case
-caseih
 caserta.it
 cash
 casino
@@ -2046,6 +2049,7 @@ ed.jp
 ed.pw
 edeka
 edgeapp.net
+edgecompute.app
 edgestack.me
 edogawa.tokyo.jp
 edu
@@ -2299,6 +2303,7 @@ eu.meteorapp.com
 eu.org
 eu.platform.sh
 eun.eg
+eurodir.ru
 eurovision
 eus
 evenassi.no
@@ -4772,9 +4777,11 @@ mc
 mc.ax
 mc.eu.org
 mc.it
+mcdir.me
 mcdir.ru
 mckinsey
 mcpe.me
+mcpre.ru
 md
 md.ci
 md.us
@@ -5553,7 +5560,6 @@ neues.museum
 neustar
 new
 newhampshire.museum
-newholland
 newjersey.museum
 newmexico.museum
 newport.museum
diff --git a/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunction.java b/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunction.java
index fae598251..69ab2cf32 100644
--- a/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunction.java
+++ b/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunction.java
@@ -63,7 +63,7 @@ import com.linecorp.armeria.server.annotation.RequestConverterFunction;
  * The {@link Parser} for JSON is applied only when the {@code content-type} of
  * the {@link RequestHeaders} is either {@link MediaType#JSON} or ends with {@code +json}.
  *
- * <h3>Conversion of multiple Protobuf messages</h3>
+ * <h2>Conversion of multiple Protobuf messages</h2>
  * A sequence of Protocol Buffer messages can not be handled by this {@link RequestConverterFunction},
  * because Protocol Buffers wire format is not self-delimiting.
  * See
diff --git a/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java b/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java
index a48e82949..1ea441a22 100644
--- a/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java
+++ b/protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java
@@ -55,7 +55,7 @@ import com.linecorp.armeria.server.annotation.ResponseConverterFunction;
  * <a href="https://developers.google.com/protocol-buffers/docs/encoding">Protocol Buffers</a> or
  * <a href="https://developers.google.com/protocol-buffers/docs/proto3#json">JSON</a> format.
  *
- * <h3>Conversion of multiple Protobuf messages</h3>
+ * <h2>Conversion of multiple Protobuf messages</h2>
  * A sequence of Protocol Buffer messages can not be handled by this {@link ResponseConverterFunction},
  * because Protocol Buffers wire format is not self-delimiting.
  * See
diff --git a/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/HealthCheckServiceConfigurator.java b/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/HealthCheckServiceConfigurator.java
index 2834ff00e..6fdc3e5cc 100644
--- a/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/HealthCheckServiceConfigurator.java
+++ b/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/HealthCheckServiceConfigurator.java
@@ -24,7 +24,7 @@ import com.linecorp.armeria.server.healthcheck.HealthCheckService;
 import com.linecorp.armeria.server.healthcheck.HealthCheckServiceBuilder;
 
 /**
- * Interface used to configure a {@link HealthCheckService} on the default armeria server.
+ * Interface used to configure a {@code HealthCheckService} on the default armeria server.
  */
 @FunctionalInterface
 public interface HealthCheckServiceConfigurator extends Ordered {
-- 
2.24.3 (Apple Git-128)

**Comment by corona10** [comment]
The above patch file might solve this issue

**Comment by minwoox** [comment]
Thanks for the diffs. 😄 
I tried this before but the Proguard we are using for the build didn't support JDK 15.
I'm not sure if it supports JDK 15 right now but it might be worth trying.

Could you try with the current Proguard, please?

**Comment by ikhoon** [comment]
Or if Proguard is a blocker, it's time to use shadow's built-in [minimizing](https://imperceptiblethoughts.com/shadow/configuration/minimizing/) and [relocation](https://imperceptiblethoughts.com/shadow/configuration/relocation/)?

**Comment by corona10** [comment]
@minwoox @ikhoon 

> Could you try with the current Proguard, please?

Apparently, even if we fix the Javadoc issue, JDK 15 can't build the Armeria fully because of the Proguard issue.
We can discuss the Proguard issue separately or update the issue name into such as `Update Armeria project to support JDK15`

**Comment by ikhoon** [comment]
There are two problems with JDK 15. One is `h3` tags in Javadoc. Another one is Proguard.
We could focus on Javadoc with this issue and probably create a separate issue on `Proguard`?

**Comment by corona10** [comment]
> We could focus on Javadoc with this issue and probably create a separate issue on Proguard?

+1

**Comment by corona10** [comment]
@ikhoon @minwoox 
Is it okay to submit the PR based on the attached patch file to close this issue?

**Comment by ikhoon** [comment]
How about upgrading Progard or migrate to shadow first?
It would be a smooth upgrade if we apply this patch and JDK 15 together.

**Comment by minwoox** [comment]
https://github.com/Guardsquare/proguard/issues/64#issuecomment-755295941
Proguard will support JDK 15 soon.

I'm also fine using Shadow. BTW, what's the benefit?

**Comment by ikhoon** [comment]
Shadow is tightly integrated with Gradle. So we can remove our custom scripts and configurations in `java-shade.gradle`.

**Comment by minwoox** [comment]
That sounds good. 😄

**Comment by ikhoon** [comment]
@corona10 This was fixed by #3426. You can now build Armeria with Java 15. 😀
